File: libgomp/testsuite/libgomp.oacc-fortran/atomic_update-1.f90
! { dg-do run }

program main
  integer igot, iexp, iexpr
  real fgot, fexp
  integer i
  integer, parameter :: N = 32
  logical lgot, lexp

  fgot = 1234.0
  fexp = 1266.0

  !$acc parallel loop copy (fgot)
    do i = 1, N
  !$acc atomic update
      fgot = fgot + 1.0
  !$acc end atomic
    end do
  !$acc end parallel loop

  if (fgot /= fexp) STOP 1

  fgot = 1.0
  fexp = 2.0**32

  !$acc parallel loop copy (fgot)
    do i = 1, N
  !$acc atomic update
      fgot = fgot * 2.0
  !$acc end atomic
    end do
  !$acc end parallel loop

  if (fgot /= fexp) STOP 2

  fgot = 32.0
  fexp = fgot - N

  !$acc parallel loop copy (fgot)
    do i = 1, N
  !$acc atomic update
      fgot = fgot - 1.0
  !$acc end atomic
    end do
  !$acc end parallel loop

  if (fgot /= fexp) STOP 3

  fgot = 2**32.0
  fexp = 1.0

  !$acc parallel loop copy (fgot)
    do i = 1, N
  !$acc atomic update
      fgot = fgot / 2.0
  !$acc end atomic
    end do
  !$acc end parallel loop

  if (fgot /= fexp) STOP 4

  lgot = .TRUE.
  lexp = .FALSE.

  !$acc parallel copy (lgot)
  !$acc atomic update
    lgot = lgot .and. .FALSE.
  !$acc end atomic
  !$acc end parallel

  if (lgot .neqv. lexp) STOP 5

  lgot = .FALSE.
  lexp = .FALSE.

  !$acc parallel copy (lgot)
  !$acc atomic update
    lgot = lgot .or. .FALSE.
  !$acc end atomic
  !$acc end parallel

  if (lgot .neqv. lexp) STOP 6

  lgot = .FALSE.
  lexp = .FALSE.

  !$acc parallel copy (lgot)
  !$acc atomic update
    lgot = lgot .eqv. .TRUE.
  !$acc end atomic
  !$acc end parallel

  if (lgot .neqv. lexp) STOP 7

  lgot = .FALSE.
  lexp = .TRUE.

  !$acc parallel copy (lgot)
  !$acc atomic update
    lgot = lgot .neqv. .TRUE.
  !$acc end atomic
  !$acc end parallel

  if (lgot .neqv. lexp) STOP 8

  fgot = 1234.0
  fexp = 1266.0

  !$acc parallel loop copy (fgot)
    do i = 1, N
  !$acc atomic update
      fgot = 1.0 + fgot
  !$acc end atomic
    end do
  !$acc end parallel loop

  if (fgot /= fexp) STOP 9

  fgot = 1.0
  fexp = 2.0**32

  !$acc parallel loop copy (fgot)
    do i = 1, N
  !$acc atomic update
      fgot = 2.0 * fgot
  !$acc end atomic
    end do
  !$acc end parallel loop

  if (fgot /= fexp) STOP 10

  fgot = 32.0
  fexp = 32.0

  !$acc parallel loop copy (fgot)
    do i = 1, N
  !$acc atomic update
      fgot = 2.0 - fgot
  !$acc end atomic
    end do
  !$acc end parallel loop

  if (fgot /= fexp) STOP 11

  fgot = 2.0**16
  fexp = 2.0**16

  !$acc parallel loop copy (fgot)
    do i = 1, N
  !$acc atomic update
      fgot = 2.0 / fgot
  !$acc end atomic
    end do
  !$acc end parallel loop

  if (fgot /= fexp) STOP 12

  lgot = .TRUE.
  lexp = .FALSE.

  !$acc parallel copy (lgot)
  !$acc atomic update
    lgot = .FALSE. .and. lgot
  !$acc end atomic
  !$acc end parallel

  if (lgot .neqv. lexp) STOP 13

  lgot = .FALSE.
  lexp = .FALSE.

  !$acc parallel copy (lgot)
  !$acc atomic update
    lgot = .FALSE. .or. lgot
  !$acc end atomic
  !$acc end parallel

  if (lgot .neqv. lexp) STOP 14

  lgot = .FALSE.
  lexp = .FALSE.

  !$acc parallel copy (lgot)
  !$acc atomic update
    lgot = .TRUE. .eqv. lgot
  !$acc end atomic
  !$acc end parallel

  if (lgot .neqv. lexp) STOP 15

  lgot = .FALSE.
  lexp = .TRUE.

  !$acc parallel copy (lgot)
  !$acc atomic update
    lgot = .TRUE. .neqv. lgot
  !$acc end atomic
  !$acc end parallel

  if (lgot .neqv. lexp) STOP 16

  igot = 1
  iexp = N

  !$acc parallel loop copy (igot)
    do i = 1, N
  !$acc atomic update
      igot = max (igot, i)
  !$acc end atomic
    end do
  !$acc end parallel loop

  if (igot /= iexp) STOP 17

  igot = N
  iexp = 1

  !$acc parallel loop copy (igot)
    do i = 1, N
  !$acc atomic update
      igot = min (igot, i)
  !$acc end atomic
    end do
  !$acc end parallel loop

  if (igot /= iexp) STOP 18

  igot = -1
  iexp = 0

  !$acc parallel loop copy (igot)
    do i = 0, N - 1
      iexpr = ibclr (-2, i)
  !$acc atomic update
      igot = iand (igot, iexpr)
  !$acc end atomic
    end do
  !$acc end parallel loop

  if (igot /= iexp) STOP 19

  igot = 0
  iexp = -1 

  !$acc parallel loop copy (igot)
    do i = 0, N - 1
      iexpr = lshift (1, i)
  !$acc atomic update
      igot = ior (igot, iexpr)
  !$acc end atomic
    end do
  !$acc end parallel loop

  if (igot /= iexp) STOP 20

  igot = -1
  iexp = 0 

  !$acc parallel loop copy (igot)
    do i = 0, N - 1
      iexpr = lshift (1, i)
  !$acc atomic update
      igot = ieor (igot, iexpr)
  !$acc end atomic
    end do
  !$acc end parallel loop

  if (igot /= iexp) STOP 21

  igot = 1
  iexp = N

  !$acc parallel loop copy (igot)
    do i = 1, N
  !$acc atomic update
      igot = max (i, igot)
  !$acc end atomic
    end do
  !$acc end parallel loop

  if (igot /= iexp) STOP 22

  igot = N
  iexp = 1

  !$acc parallel loop copy (igot)
    do i = 1, N
  !$acc atomic update
      igot = min (i, igot)
  !$acc end atomic
    end do
  !$acc end parallel loop

  if (igot /= iexp) STOP 23

  igot = -1
  iexp = 0

  !$acc parallel loop copy (igot)
    do i = 0, N - 1
      iexpr = ibclr (-2, i)
  !$acc atomic update
      igot = iand (iexpr, igot)
  !$acc end atomic
    end do
  !$acc end parallel loop

  if (igot /= iexp) STOP 24

  igot = 0
  iexp = -1 

  !$acc parallel loop copy (igot)
    do i = 0, N - 1
        iexpr = lshift (1, i)
  !$acc atomic update
      igot = ior (iexpr, igot)
  !$acc end atomic
    end do
  !$acc end parallel loop

  if (igot /= iexp) STOP 25

  igot = -1
  iexp = 0 

  !$acc parallel loop copy (igot)
    do i = 0, N - 1
      iexpr = lshift (1, i)
  !$acc atomic update
      igot = ieor (iexpr, igot)
  !$acc end atomic
    end do
  !$acc end parallel loop

  if (igot /= iexp) STOP 26

end program

File: libgomp/testsuite/libgomp.oacc-c-c++-common/atomic_update-1.c
/* { dg-do run } */

#include <stdlib.h>

int
main(int argc, char **argv)
{
  float fexp, fgot;
  int   iexp, igot;
  long long lexp, lgot;
  int   N = 32;
  int	i;

  fgot = 1234.0;
  fexp = 1235.0;

#pragma acc data copy (fgot)
  {
#pragma acc parallel loop
    for (i = 0; i < 1; i++)
#pragma acc atomic update
      fgot++;
  }

  if (fexp != fgot)
    abort ();

  fgot = 1234.0;
  fexp = fgot - N;

#pragma acc data copy (fgot)
  {
#pragma acc parallel loop
    for (i = 0; i < N; i++)
      {
#pragma acc atomic update
        fgot--;
      }
  }

  if (fexp != fgot)
    abort ();

  fgot = 1234.0;
  fexp = fgot + N;

#pragma acc data copy (fgot)
  {
#pragma acc parallel loop
    for (i = 0; i < N; i++)
      {
#pragma acc atomic update
        ++fgot;
      }
  }

  if (fexp != fgot)
    abort ();

  fgot = 1234.0;
  fexp = fgot - N;

#pragma acc data copy (fgot)
  {
#pragma acc parallel loop
    for (i = 0; i < N; i++)
      {
#pragma acc atomic update
        --fgot;
      }
  }

  if (fexp != fgot)
    abort ();

  /* BINOP = + */

  fgot = 1234.0;
  fexp = fgot + N;

#pragma acc data copy (fgot)
  {
#pragma acc parallel loop
    for (i = 0; i < N; i++)
      {
        float expr = 1.0;
#pragma acc atomic update
        fgot += expr;
      }
  }

  if (fexp != fgot)
    abort ();

  fgot = 1234.0;
  fexp = fgot + N;

#pragma acc data copy (fgot)
  {
#pragma acc parallel loop
    for (i = 0; i < N; i++)
      {
        float expr = 1.0;
#pragma acc atomic update
        fgot = fgot + expr;
      }
  }

  if (fexp != fgot)
    abort ();

  fgot = 1234.0;
  fexp = fgot + N;

#pragma acc data copy (fgot)
  {
#pragma acc parallel loop
    for (i = 0; i < N; i++)
      {
        float expr = 1.0;
#pragma acc atomic update
        fgot = expr + fgot;
      }
  }

  if (fexp != fgot)
    abort ();

  fgot = 1234.0;

#pragma acc data copy (fgot)
  {
#pragma acc parallel loop
    for (i = 0; i < N; i++)
      {
        float expr = 0.5;
#pragma acc atomic update
        fgot = (expr + expr) + fgot;
      }
  }

  if (fexp != fgot)
    abort ();

  /* BINOP = * */

  fgot = 1234.0;
  fexp = 1234.0;

  for (i = 0; i < N; i++)
    fexp *= 2.0;

#pragma acc data copy (fgot)
  {
#pragma acc parallel loop
    for (i = 0; i < N; i++)
      {
        float expr = 2.0;
#pragma acc atomic update
        fgot *= expr;
      }
  }

  if (fexp != fgot)
    abort ();

  fgot = 1234.0;
  fexp = 1234.0;

  for (i = 0; i < N; i++)
    fexp = fexp * 2.0;

#pragma acc data copy (fgot)
  {
#pragma acc parallel loop
    for (i = 0; i < N; i++)
      {
        float expr = 2.0;
#pragma acc atomic update
        fgot = fgot * expr;
      }
  }

  if (fexp != fgot)
    abort ();

  fgot = 1234.0;
  fexp = 1234.0;

  for (i = 0; i < N; i++)
    fexp = 2.0 * fexp;

#pragma acc data copy (fgot)
  {
#pragma acc parallel loop
    for (i = 0; i < N; i++)
      {
        float expr = 2.0;
#pragma acc atomic update
        fgot = expr * fgot;
      }
  }

  if (fexp != fgot)
    abort ();

  fgot = 1234.0;

#pragma acc data copy (fgot)
  {
#pragma acc parallel loop
    for (i = 0; i < N; i++)
      {
        float expr = 1.0;
#pragma acc atomic update
        fgot = (expr + expr) * fgot;
      }
  }

  if (fexp != fgot)
    abort ();

  /* BINOP = - */

  fgot = 1234.0;
  fexp = 1234.0;

  for (i = 0; i < N; i++)
    fexp -= 2.0;

#pragma acc data copy (fgot)
  {
#pragma acc parallel loop
    for (i = 0; i < N; i++)
      {
        float expr = 2.0;
#pragma acc atomic update
        fgot -= expr;
      }
  }

  if (fexp != fgot)
    abort ();

  fgot = 1234.0;
  fexp = 1234.0;

  for (i = 0; i < N; i++)
    fexp = fexp - 2.0;

#pragma acc data copy (fgot)
  {
#pragma acc parallel loop
    for (i = 0; i < N; i++)
      {
        float expr = 2.0;
#pragma acc atomic update
        fgot = fgot - expr;
      }
  }

  if (fexp != fgot)
    abort ();

  fgot = 1234.0;
  fexp = 1234.0;

  for (i = 0; i < N; i++)
    fexp = 2.0 - fexp;

#pragma acc data copy (fgot)
  {
#pragma acc parallel loop
    for (i = 0; i < N; i++)
      {
        float expr = 2.0;

#pragma acc atomic update
        fgot = expr - fgot;
      }
  }

  if (fexp != fgot)
    abort ();

  fgot = 1234.0;

#pragma acc data copy (fgot)
  {
#pragma acc parallel loop
    for (i = 0; i < N; i++)
      {
        float expr = 1.0;
#pragma acc atomic update
        fgot = (expr + expr) - fgot;
      }
  }

  if (fexp != fgot)
    abort ();

  /* BINOP = / */

  fgot = 1234.0;
  fexp = 1234.0;

  for (i = 0; i < N; i++)
    fexp /= 2.0;

#pragma acc data copy (fgot)
  {
#pragma acc parallel loop
    for (i = 0; i < N; i++)
      {
        float expr = 2.0;
#pragma acc atomic update
        fgot /= expr;
      }
  }

  if (fexp != fgot)
    abort ();

  fgot = 1234.0;
  fexp = 1234.0;

  for (i = 0; i < N; i++)
    fexp = fexp / 2.0;

#pragma acc data copy (fgot)
  {
#pragma acc parallel loop
    for (i = 0; i < N; i++)
      {
        float expr = 2.0;
	
#pragma acc atomic update
        fgot = fgot / expr;
      }
  }

  if (fexp != fgot)
    abort ();

  fgot = 1234.0;
  fexp = 1234.0;

  for (i = 0; i < N; i++)
    fexp = 2.0 / fexp;

#pragma acc data copy (fgot)
  {
#pragma acc parallel loop
    for (i = 0; i < N; i++)
      {
        float expr = 2.0;

#pragma acc atomic update
        fgot = expr / fgot;
      }
  }

  if (fexp != fgot)
    abort ();

  fgot = 1234.0;
  fexp = 1234.0;

  for (i = 0; i < N; i++)
    fexp = 2.0 / fexp;

#pragma acc data copy (fgot)
  {
#pragma acc parallel loop
    for (i = 0; i < N; i++)
      {
        float expr = 1.0;
#pragma acc atomic update
        fgot = (expr + expr) / fgot;
      }
  }

  if (fexp != fgot)
    abort ();

  /* BINOP = & */

  igot = ~0;
  iexp = 0;

#pragma acc data copy (igot)
  {
#pragma acc parallel loop
    for (i = 0; i < N; i++)
      {
        int expr = ~(1 << i);

#pragma acc atomic update
        igot &= expr;
      }
  }

  if (iexp != igot)
    abort ();

  igot = ~0;
  iexp = 0;

#pragma acc data copy (igot)
  {
#pragma acc parallel loop
    for (i = 0; i < N; i++)
      {
        int expr = ~(1 << i);
#pragma acc atomic update
        igot = igot / expr;
      }
  }

  if (iexp != igot)
    abort ();

  igot = ~0;
  iexp = 0;

#pragma acc data copy (igot)
  {
#pragma acc parallel loop
    for (i = 0; i < N; i++)
      {
        int expr = ~(1 << i);
#pragma acc atomic update
        igot = expr & igot;
     }
  }

  if (iexp != igot)
    abort ();

  igot = ~0;
  iexp = 0;

#pragma acc data copy (igot)
  {
#pragma acc parallel loop
    for (i = 0; i < N; i++)
      {
        int expr = ~(1 << i);
        int zero = 0;

#pragma acc atomic update
        igot = (expr + zero) & igot;
      }
  }

  if (iexp != igot)
    abort ();

  /* BINOP = ^ */

  igot = ~0;
  iexp = 0;

#pragma acc data copy (igot)
  {
#pragma acc parallel loop
    for (i = 0; i < N; i++)
      {
        int expr = (1 << i);

#pragma acc atomic update
        igot ^= expr;
      }
  }

  if (iexp != igot)
    abort ();

  igot = ~0;
  iexp = 0;

#pragma acc data copy (igot)
  {
#pragma acc parallel loop
    for (i = 0; i < N; i++)
      {
        int expr = (1 << i);

#pragma acc atomic update
        igot = igot ^ expr;
      }
  }

  if (iexp != igot)
    abort ();

  igot = ~0;
  iexp = 0;

#pragma acc data copy (igot)
  {
#pragma acc parallel loop
    for (i = 0; i < N; i++)
      {
        int expr = (1 << i);

#pragma acc atomic update
        igot = expr ^ igot;
      }
  }

  if (iexp != igot)
    abort ();

  igot = ~0;
  iexp = 0;

#pragma acc data copy (igot)
  {
#pragma acc parallel loop
    for (i = 0; i < N; i++)
      {
        int expr = (1 << i);
        int zero = 0;

#pragma acc atomic update
        igot = (expr + zero) ^ igot;
      }
  }

  if (iexp != igot)
    abort ();

  /* BINOP = | */

  igot = 0;
  iexp = ~0;

#pragma acc data copy (igot)
  {
#pragma acc parallel loop
    for (i = 0; i < N; i++)
      {
        int expr = (1 << i);

#pragma acc atomic update
        igot |= expr;
      }
  }

  if (iexp != igot)
    abort ();

  igot = 0;
  iexp = ~0;

#pragma acc data copy (igot)
  {
#pragma acc parallel loop
    for (i = 0; i < N; i++)
      {
        int expr = (1 << i);

#pragma acc atomic update
        igot = igot | expr;
      }
  }

  if (iexp != igot)
    abort ();

  igot = 0;
  iexp = ~0;

#pragma acc data copy (igot)
  {
#pragma acc parallel loop
    for (i = 0; i < N; i++)
      {
        int expr = (1 << i);

#pragma acc atomic update
        igot = expr | igot;
      }
  }

  if (iexp != igot)
    abort ();

  igot = 0;
  iexp = ~0;

#pragma acc data copy (igot)
  {
#pragma acc parallel loop
    for (i = 0; i < N; i++)
      {
        int expr = (1 << i);
        int zero = 0;

#pragma acc atomic update
        igot = (expr + zero) | igot;
      }
  }

  if (iexp != igot)
    abort ();

  /* BINOP = << */

  lgot = 1LL;
  lexp = 1LL << N;

#pragma acc data copy (lgot)
  {
#pragma acc parallel loop
    for (i = 0; i < N; i++)
      {
        long long expr = 1LL;

#pragma acc atomic update
        lgot <<= expr;
      }
  }

  if (lexp != lgot)
    abort ();

  lgot = 1LL;
  lexp = 1LL << N;

#pragma acc data copy (lgot)
  {
#pragma acc parallel loop
    for (i = 0; i < N; i++)
      {
        long long expr = 1LL;

#pragma acc atomic update
        lgot = lgot << expr;
      }
  }

  if (lexp != lgot)
    abort ();

  lgot = 1LL;
  lexp = 2LL;

#pragma acc data copy (lgot)
  {
#pragma acc parallel loop
    for (i = 0; i < 1; i++)
      {
        long long expr = 1LL;

#pragma acc atomic update
        lgot = expr << lgot;
      }
  }

  if (lexp != lgot)
    abort ();

  lgot = 1LL;
  lexp = 2LL;

#pragma acc data copy (lgot)
  {
#pragma acc parallel loop
    for (i = 0; i < 1; i++)
      {
        long long expr = 1LL;
        long long zero = 0LL;

#pragma acc atomic update
        lgot = (expr + zero) << lgot;
      }
  }

  if (lexp != lgot)
    abort ();

  /* BINOP = >> */

  lgot = 1LL << N;
  lexp = 1LL;

#pragma acc data copy (lgot)
  {
#pragma acc parallel loop
    for (i = 0; i < N; i++)
      {
        long long expr = 1LL;

#pragma acc atomic update
        lgot >>= expr;
      }
  }

  if (lexp != lgot)
    abort ();

  lgot = 1LL << N;
  lexp = 1LL;

#pragma acc data copy (lgot)
  {
#pragma acc parallel loop
    for (i = 0; i < N; i++)
      {
        long long expr = 1LL;

#pragma acc atomic update
        lgot = lgot >> expr;
      }
  }

  if (lexp != lgot)
    abort ();

  lgot = 1LL;
  lexp = 1LL << (N - 1);

#pragma acc data copy (lgot)
  {
#pragma acc parallel loop
    for (i = 0; i < 1; i++)
      {
        long long expr = 1LL << N;

#pragma acc atomic update
        lgot = expr >> lgot;
    }
  }

  if (lexp != lgot)
    abort ();

  lgot = 1LL;
  lexp = 1LL << (N - 1);

#pragma acc data copy (lgot)
  {
#pragma acc parallel loop
    for (i = 0; i < 1; i++)
      {
        long long expr = 1LL << N;
        long long zero = 0LL;

#pragma acc atomic update
        lgot = (expr + zero) >> lgot;
    }
  }

  if (lexp != lgot)
    abort ();

  return 0;
}
